include/oneapi/dpl/pstl/zip_view_impl.h

```diff
@@ -0,0 +1,548 @@
+            __internal::__all_random_access<_Const, _Views...>, std::random_access_iterator_tag,
+            std::conditional_t<__internal::__all_bidirectional<_Const, _Views...>, std::bidirectional_iterator_tag,
+                               std::conditional_t<__internal::__all_forward<_Const, _Views...>,
+                                                  std::forward_iterator_tag, std::input_iterator_tag>>>;
+
+        using value_type = __tuple_type<std::ranges::range_value_t<__internal::__maybe_const<_Const, _Views>>...>;
+        using difference_type =
+            std::common_type_t<std::ranges::range_difference_t<__internal::__maybe_const<_Const, _Views>>...>;
+
+      private:
+        using __reference_type =
+            __tuple_type<std::ranges::range_reference_t<__internal::__maybe_const<_Const, _Views>>...>;
+        using __rvalue_reference_type =
+            __tuple_type<std::ranges::range_rvalue_reference_t<__internal::__maybe_const<_Const, _Views>>...>;
+
+        using __iterators_type = __tuple_type<std::ranges::iterator_t<__internal::__maybe_const<_Const, _Views>>...>;
+
+      public:
+        iterator() = default;
+        constexpr iterator(iterator<!_Const> i)
+            requires _Const &&
+                     (std::convertible_to<std::ranges::iterator_t<_Views>,
+                                          std::ranges::iterator_t<__internal::__maybe_const<_Const, _Views>>> &&
+                      ...)
+            : __current(std::move(i.__current))
+        {
+        }
+
+      private:
+        constexpr explicit iterator(__iterators_type __current) : __current(std::move(__current)) {}
+
+      public:
+        template <typename... Iterators>
+        operator oneapi::dpl::zip_iterator<Iterators...>() const
+        {
+            auto __tr = [](auto&&... __args) -> decltype(auto) {
+                return oneapi::dpl::make_zip_iterator(std::forward<decltype(__args)>(__args)...);
+            };
+            return __internal::__apply_to_tuple([](auto it) -> decltype(auto) { return it; }, __current, __tr);
+        }
+
+        constexpr decltype(auto)
+        operator*() const
+        {
+            auto __tr = [](auto&&... __args) -> decltype(auto) {
+                return __reference_type(std::forward<decltype(__args)>(__args)...);
+            };
+            return __internal::__apply_to_tuple([](auto& __it) -> decltype(auto) { return *__it; }, __current, __tr);
+        }
+
+        constexpr decltype(auto)
+        operator[](difference_type __n) const
+            requires __internal::__all_random_access<_Const, _Views...>
+        {
+            return *(*this + __n);
+        }
+
+        constexpr iterator&
+        operator++()
+        {
+            __internal::__apply_to_tuple([](auto& __it) -> decltype(auto) { return ++__it; }, __current);
+            return *this;
+        }
+
+        constexpr void
+        operator++(int)
+        {
+            ++*this;
+        }
```

Comment:
I don't understand what this function is here for (and how it doesn't conflict with the subsequent function).

Reply:
This overload is part of the [zip_view specification](https://en.cppreference.com/w/cpp/ranges/zip_view/iterator/operator_arith).

Practically, it works only for the pure input iterators. As far as I understand, it was defined this way because for pure input iterator, returning the copy of the previous state of iterator does not make many sense because the copy is invalidated after doing `++*this`.

It does not conflict with the next overload because the next one has some constraints defined and for forward iterators and above, the next function would be considered as the more constrained overload and is guaranteed to be selected.

Reply:
Understood.  Thanks for the explanation.

Reply:
Probably, it juts for input iterators support...
There is a test case:

```
// all input+
    int buffer[3] = {4, 5, 6};
    dpl_ranges::zip_view v(a, InputRange{buffer});
    auto it = v.begin();
    using Iter = decltype(it);

    assert(&(std::get<0>(*it)) == &(a[0]));
    assert(&(std::get<1>(*it)) == &(buffer[0]));

    static_assert(std::is_same_v<decltype(++it), Iter&>);
    auto& it_ref = ++it;
    assert(&it_ref == &it);
    assert(&(std::get<0>(*it)) == &(a[1]));
    assert(&(std::get<1>(*it)) == &(buffer[1]));

    static_assert(std::is_same_v<decltype(it++), void>);
    it++;
    assert(&(std::get<0>(*it)) == &(a[2]));
    assert(&(std::get<1>(*it)) == &(buffer[2]));
```